src/cpp/ext/filters/logging/logging_filter.cc

```diff
@@ -0,0 +1,503 @@
+//
+//
+// Copyright 2022 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/cpp/ext/filters/logging/logging_filter.h"
+
+#include <cstddef>
+#include <limits>
+
+#include "absl/random/distributions.h"
+#include "absl/random/random.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/str_split.h"
+#include "logging_sink.h"
+
+#include "src/core/ext/filters/client_channel/client_channel.h"
+#include "src/core/lib/channel/promise_based_filter.h"
+#include "src/core/lib/config/core_configuration.h"
+#include "src/core/lib/gprpp/host_port.h"
+#include "src/core/lib/promise/cancel_callback.h"
+#include "src/core/lib/promise/map_pipe.h"
+#include "src/core/lib/promise/promise.h"
+#include "src/core/lib/promise/seq.h"
+#include "src/core/lib/promise/try_concurrently.h"
+#include "src/cpp/ext/filters/logging/logging_sink.h"
+
+namespace grpc {
+namespace internal {
+
+namespace {
+
+LoggingSink* g_logging_sink = nullptr;
+
+uint64_t GetCallId() {
+  thread_local absl::InsecureBitGen gen;
+  return absl::Uniform(gen, 0u, std::numeric_limits<uint64_t>::max());
+}
+
+class LoggingFilter : public grpc_core::ChannelFilter {
+ protected:
+  explicit LoggingFilter() {}
+
+ private:
+};
+
+class MetadataEncoder {
+ public:
+  MetadataEncoder(LoggingSink::Entry::Payload* payload,
+                  absl::string_view* status_details_bin, uint64_t log_len)
+      : payload_(payload),
+        status_details_bin_(status_details_bin),
+        log_len_(log_len) {}
+
+  void Encode(const grpc_core::Slice& key_slice,
+              const grpc_core::Slice& value_slice) {
+    auto key = key_slice.as_string_view();
+    auto value = value_slice.as_string_view();
+    if (status_details_bin_ != nullptr && key == "grpc-status-details-bin") {
+      *status_details_bin_ = value;
+      return;
+    }
+    if (absl::ConsumePrefix(&key, "grpc-")) {
+      // skip all other grpc- headers
+      return;
+    }
+    uint64_t mdentry_len = key.length() + value.length();
+    if (mdentry_len > log_len_) {
+      gpr_log(GPR_DEBUG,
```

Comment:
It'd be useful to say mat's max and what's current value in the log message.

Reply:
Done